lib/LanguageServerRename/Handler/RenameHandler.php

```diff
@@ -0,0 +1,124 @@
+use Phpactor\LanguageServerProtocol\RenameRequest;
+use Phpactor\LanguageServerProtocol\ServerCapabilities;
+use Phpactor\LanguageServerProtocol\TextDocumentEdit;
+use Phpactor\LanguageServerProtocol\VersionedTextDocumentIdentifier;
+use Phpactor\LanguageServerProtocol\WorkspaceEdit;
+use Phpactor\LanguageServer\Core\Handler\CanRegisterCapabilities;
+use Phpactor\LanguageServer\Core\Handler\Handler;
+use Phpactor\LanguageServer\Core\Workspace\Workspace;
+use Phpactor\TextDocument\TextDocumentLocator;
+use Phpactor\TextDocument\TextDocumentUri;
+
+class RenameHandler implements Handler, CanRegisterCapabilities
+{
+    /** @var Renamer */
+    private $renamer;
+    /** @var Workspace */
+    private $workspace;
+    /** @var TextDocumentLocator */
+    private $documentLocator;
+
+    public function __construct(Workspace $workspace, TextDocumentLocator $documentLocator, Renamer $renamer)
+    {
+        $this->renamer = $renamer;
+        $this->workspace = $workspace;
+        $this->documentLocator = $documentLocator;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function methods(): array
+    {
+        return [
+            PrepareRenameRequest::METHOD => 'prepareRename',
+            RenameRequest::METHOD => 'rename',
+        ];
+    }
+    /** @return Promise<WorkspaceEdit> */
+    public function rename(RenameParams $params): Promise
+    {
+        return \Amp\call(function () use ($params) {
+            return $this->resultToWorkspaceEdit(
+                $this->renamer->rename(
+                    $document = $this->documentLocator->get(TextDocumentUri::fromString($params->textDocument->uri)),
+                    PositionConverter::positionToByteOffset(
+                        $params->position,
+                        (string)$document
+                    ),
+                    $params->newName
+                )
+            );
+        });
+    }
+    /** @return Promise<Range> */
+    public function prepareRename(PrepareRenameParams $params): Promise
+    {
+        // https://microsoft.github.io/language-server-protocol/specification#textDocument_prepareRename
+        return \Amp\call(function () use ($params) {
+            $range = $this->renamer->getRenameRange(
+                $document = $this->documentLocator->get(TextDocumentUri::fromString($params->textDocument->uri)),
+                PositionConverter::positionToByteOffset(
+                    $params->position,
+                    (string)$document
+                ),
+            );
+            if ($range == null) {
+                return null;
+            }
+            return new Range(
+                PositionConverter::byteOffsetToPosition(
+                    $range->start(),
+                    (string)$document
+                ),
+                PositionConverter::byteOffsetToPosition(
+                    $range->end(),
+                    (string)$document
+                )
+            );
```

Comment:
Could probably add a utility to convert a Phpactor Range => an LSP Range in the LanguageServerBridge

Reply:
Agreed. Was thinking about that.

Reply:
Do you want _me_ to do that or will you do it?

Reply:
Up to you, feel free to leave it.

Reply:
Yea, I probably should do it then....